### DISCUSSION TITLE
Duplicated Folders

### DISCUSSION THREAD

**Discussion by Node815** [dani-garcia/vaultwarden#3640] [Q&A] [answered]
I'm not sure how this came about, but I did a reinstall to a new OS for the server and imported my vault from a fresh install BTW, and now I have duplicate folders with identical contents between both.   Aside from exporting the vault, editing it in a spreadsheet program to remove the duplicates en masse' and, "nuking" the vault, then re-importing it, is there a better way of doing this than hitting the folder via the web panel and selecting edit>delete for each one at a time? 

I'm not adverse to exporting and doing the above, but I was hoping there was an easier/quicker way.

**Comment by tessus** [comment, marked as answer]
I guess the following happened:

- you exported your vault (from web-vault)
- you are using the web-vault and maybe another client (which you didn't wipe)
- you set up the new server and imported your vault (to your empty web-vault)
- the other client syncs (uploads) the entries

I don't know why, but for some reason you can import the same set of data several times and it will always create new entries.

IMO this is a bug because the exported data has the item ids and the folder ids, which should be the primary keys in the database. When an import happens, there should be a constraint violation and nothing should be imported.

Either way, the solution is: make sure you have a proper export and wipe everything (all devices and web-vault). Import your data once.

I don't know what would have happened, if you haden't imported anything. I think the device with the most recent data would have synced the data to the server. But that's just a guess. I don't know the exact sync protocol. Something you can test, if you are bored.

**Comment by BlackDex** [comment]
The old clients couldn't have synced, because they need to login again because there device information wasn't available because it was a full new install. Which means no refresh tokens or login tokens which were valid.

It depends on how you export i think, via json it should dedup as @tessus mentioned. Not sure why it doesn't.

**Reply by tessus** [reply]
> via json it should dedup as @tessus mentioned. Not sure why it doesn't.

I tried with json. You can try it on a test instance. export/import - duplicated. import again - 3 sets of the same credentials. and it goes on and on...

P.S.: I tried with 19e671ff25bffa47424b5af44264c2c74c2cc84b and web 2023.05.0

**Reply by BlackDex** [reply]
I actually meant, that from a programmers mindset it should. But apparently it doesn't.

**Reply by tessus** [reply]
I don't have a bitwarden account, thus I don't know. Is this a
- bitwarden design issue
- vaultwarden server bug
- a combination of a client issue with one of the above

**Reply by tessus** [reply]
@BlackDex do you know the answer to ^^^?

**Reply by BlackDex** [reply]
I didn't have the time yet to check and verify.

**Reply by tessus** [reply]
No worries, I just thought you missed the comment. If this happens on bw server as well, we should open a ticket with them. This is a very strange situation that goes against any sane and sound design.

**Reply by BlackDex** [reply]
Well i can see it two ways. While UUIDv4 should be unique in most situations, a collision could occur. This could of course be the reason they do not use that during import. But in my opinion, you could of course match multiple items and skip if it is the same.

**Reply by tessus** [reply]
Hmm, I think I looked that up a few years back. The chances of UUIDs to collide are so low that we won't have to worry for a few billion years or so. I can't recall the exact number.
Also, how is this solved when the item is created? A new UUID is generated and then what? What if it is alraedy in the database? I would suspect that if that UUID already exists a new one is generated.
This logic is fine for creating an item, but not for importing one.

**Reply by BlackDex** [reply]
It looks like Bitwarden does exactly the same. So it's not something we can fix on our side.
We also can't match the folder names or stuff like that, since those are encrypted and we have no idea what the contents is.

So, i suggest to either report this as an issue upstream at Bitwarden, or live with it :)

**Reply by tessus** [reply]
I can try to open an issue with bw (if I can without having a bw account - who knows what info they want in their tickets). However, I am afraid they'll just close it. They are not that great with design and close even the most logical and proper issues.

I'm afraid we'll have to live with it in the end.
